Includes several changes to improve flow (#13)
hablapps/pairstrading: PostPairsTrading.md

> # A Match Made in Trading: Step-by-Step Pairs Trading Guide

Reviewer: What do you think of this title? ** Implementing Pairs Trading in Kdb+ with a Tick Architecture **
I know it's not very fancy, but at least it's descriptive :)

Author: It's fine but perhaps too obvious, I think we have to come up with something more attractive.

KDB+/Q stands out as **a powerful tool in finance**, renowned for its ability to handle vast volumes of real-time data amidst the relentless dynamics of the market. In this article, we embark on an insightful exploration of _Pairs Trading_, one of the most popular strategies in the trading world, and its implementation in Q.

> In order to achieve this exploration, we have outlined the following steps:

Reviewer 2: Not sure about the wording here. I suggest:

```suggestion
In order to achieve this, we have outlined the following steps:
```

* Identifying cointegrated indexes, i.e., related pairs
* Implementing a model to calculate their spreads
* Visualizing the approach in real-time

> We will use the [_Tick Architecture_](https://github.com/KxSystems/kdb-tick), the typical setup found in kdb+ systems, to introduce, contextualize and guide each step. As we acknowledge that the typical reader is a quant interested in learning the virtues of q, a concise introduction to this architecture will be provided. We have also aimed to briefly introduce pairs trading so that other developers can follow and benefit from this text. Please, feel free to skip the following section if you are familiar with it.

Reviewer 2: Is it KDB, kdb or Kdb? On the same note, is it Q or q? Very minor but this should be consistent

Author: I always use _kdb+_ and _q_, I think this is the most standard way of referring to them, taking into account Whitney's convention, but you're right, I'll adapt to the original style from the text: _KDB+_ and _Q_.

> 💡 Which should not be confused with correlation; cointegration is a statistical property of two-time series, indicating a long-term relationship between them despite short-term fluctuations. Cointegrated series move together over time, sharing a common stochastic drift. On the other hand, correlation measures the strength and direction of the linear relationship between two variables at a specific point in time. While correlation captures the degree of association between variables, cointegration reflects a deeper, underlying relationship that persists over time.

> Hence, we're interested in **cointegrated assets**, which are assets that exhibit the following characteristics:

Reviewer: Is this section about Pair Trading or Cointegration? Let me explain. I think if we want to introduce Pair Trading as a general strategy, it might be enough to mention that we are looking for two indices that have a relationship, and then in the chapter on Cointegration, we can explain this in more detail. So in this section, I would focus on explaining that the idea is to look for when these indices, which have a relationship, start to diverge in order to arbitrage

Author: This is a very interesting thing. I thought they were kind of synonyms, but I completely agree that we should keep things as simple as possible here and possibly skip cointegration until the proper section.

* They have a similar trend, meaning the difference between both assets maintains a constant mean, and this difference fluctuates around that same mean.
* This inherent relationship persists in the long run, meaning that our series is not dependent on time.

> Before delving into this search for cointegrated pairs, let me introduce you the tick architecture.

Reviewer 2: ```suggestion

Before diving into this search for cointegrated pairs, let me introduce you the tick architecture.
```

> ## What is the Tick Architecture?

Reviewer 2: I think that this section comes a bit out of nowhere. Maybe we should mention that we are going to use the Tick Architecture in the introduction?

Author: There is already a paragraph that refers to the tick architecture in the introduction, isn't it?

Reviewer 2: You are right. I guess I missed it

> The tick architecture is designed to handle high-frequency trading data efficiently. At its heart is the tickerplant (TP), a crucial component responsible for receiving and timestamping incoming data, then broadcasting it to other components such as the real-time database (RDB) and historical database (HDB). The TP ensures that data is distributed in a timely manner, allowing for real-time analytics and decision-making. The RDB stores recent data for quick access, while the HDB archives older data for long-term storage and analysis. Additionally, the feedhandler plays a vital role by interfacing with external data sources, making sure that the TP receives accurate and up-to-date information. This architecture guarantees seamless data flow and rapid access to both real-time and historical data, making it ideal for high-frequency trading applications.

Reviewer 2: Maybe we should explain (in a single line perhaps) what it _is_ at a higher level and how it fits our "project" before going deep into its components

Author: I think I just copied your intro to the tick architecture, but I find it ok if you want to extend it. We don't talk much about the performance and the general benefits, perhaps it would be nice to briefly discuss it right here.

![Architecture](resources/general-architecture.png)

> To develop the pair trading strategy, we have added a couple of new components to the default vanilla architecture: ADF Test, Real-time Pair Trading (RPT) and KX Dashboard. They will be introduced as required. So finally, let's move on to the very first step of this journey: finding cointegrated pairs.

Reviewer 2: ```suggestion
To develop the pair trading strategy, we have added a couple of new components to the default vanilla architecture: ADF Test, Real-time Pair Trading (RPT) and the KX Dashboard. They will be introduced when required. So finally, let's move on to the very first step of this journey: finding cointegrated pairs.

```

Author: Not sure about this one. Should we place _the_ before a proper noun? I know we are referring to a component though... Also, I think that "introduced as required" is better in this particular case.

Reviewer 2: My suggestions here had very low impact. Feel free to leave it as it was

## Identifying cointegrated indexes

> This section focuses on the ADF Test component. It reads data from the HDB and provides the means for the quant to determine a cointegrated pair as output, which will then be supplied as input for further steps.

Reviewer 2: ```suggestion
Let's focus on the ADF Test component. It reads data from the HDB and provides the means for the quant to determine a cointegrated pair as output, which will then be supplied as input for further steps.

```

![Arch 1st Part](resources/general-architecture-top.png)

> Let's move on to the code we need to implement it.

Reviewer 2: ```suggestion

Let's move on to the code we need to implement it.

```

> Imagine **we selected 13 world indexes** and aimed to assess whether they are **cointegrated** among them or not. In KDB+/Q we can start by declaring a variable containing every index and generating the cartesian product (`cross`) of this indexes.

Reviewer 2: Again, KDB, kdb or Kdb? q or Q?

```q
system"l pykx.q"
```
One such library is **statsmodels**, a prominent tool in Python for statistical modeling and hypothesis testing. It equips analysts with a robust toolkit for regression, time series, and multivariate analysis. Specifically, within the statsmodels package, the _statsmodels.tsa.stattools_ module features the Cointegration test.
```q
co:.pykx.import[`statsmodels.tsa.stattools]`:coint
```
We can proceed to create a function called `cof` to call our imported function from PyKX and return the second element from the resulting list, which in this case is the p-value, [as described in the official documentation](https://www.statsmodels.org/stable/generated/statsmodels.tsa.stattools.coint.html).
```q

> cof:@[;1]co[<]::

Reviewer: I think this is a good moment to introduce or comment on tacit programming. I propose:

As you can see, this function definition doesn't make explicit reference to its arguments. This is achieved thanks to how KDB+/q allows us to compose functions. This style of programming, where function arguments are not explicitly named, is known as tacit programming or point-free style. In KDB+/q, tacit programming is particularly powerful and concise, allowing us to create expressive and efficient code. It leverages the language's capabilities for function composition and implicit argument passing, resulting in elegant solutions like the one we've just seen.

Author: I like it. I think it would fit nicely as a quoted hint, what do you think? Also, I'd supply a link as a reference.

fcoint:{@[;1]0f^coint[x;y]`}
```

As we saw in the introduction, we are working in a tick architecture environment. This means that, in addition to receiving real-time prices for our indices, this architecture provides a tool to store the closing prices of our indices in our HDB at the end of the day. As obvious, establishing a connection with HDB is required:
```q
hdb:hopen port
```

> As can be seen, we assume that the HDB process is listening at a given `port` in the local machine. We just `hopen` a connection with it and get the `hdb` handle.

Reviewer: Here we lose the consistency we have of explaining the code first and then showing it, because here we are showing it and then explaining it. I'm not sure if it's important, just mentioning it :)

Author: I never asked for such constraint. I mean, one thing is showing several large paragraphs and later show the code at the end (which could be fine for certain situations). But here, the idea is to explain one fragment of code in each paragraph, or at least explain it little by little. It's ok if you just introduce it, show the code and right after that go deeper in certain details, while remaining in the very same paragraph.

> Then, we define the query that we want the HDB to run. In this case, we declare `rs`, which takes the (n)umber of historic days and the involved (symb)ol(s) for which we want to retrieve data, as arguments.

Reviewer 2: ```suggestion
Then, we define the query that we want the HDB to run. In this case, we declare `rs`, which takes the (n)umber of historic days and the involved (sym)bol(s) for which we want to retrieve data, as arguments.

```

```q
rs:{[tr;syms]select from prices where date within (.z.d-tr;.z.d),sym in syms}
```

> The body of the function might seem pretty familiar to the SQL practicioner. In fact, we are exploiting **qSQL syntax** here, which leverages a syntax similar to SQL but optimised for kdb+. It might also interesting to say that `.z.d` represents the current date, so we are interested on the `n` days back from today.

Reviewer: ... interested on the `tr` days back from today

Author: I think `n` fits better here, since it's not a range, it's a number of days after my requested change. I'll adapt the text accordingly.

> The body of the function might seem pretty familiar to the SQL practicioner. In fact, we are exploiting **qSQL syntax** here, which leverages a syntax similar to SQL but optimised for kdb+. It might also interesting to say that `.z.d` represents the current date, so we are interested on the `n` days back from today.

Reviewer 2: ```suggestion
The body of the function might seem pretty familiar to the SQL practitioner. In fact, we are exploiting **qSQL syntax** here, which leverages a syntax similar to SQL but optimised for kdb+. It might also interesting to say that `.z.d` represents the current date, so we are interested on the `n` days back from today.

```

Author: I think _practicioner_ is ok in this context, _practitioner_ is the doctor.

Reviewer 2: I meant practitioner as in "AWS Cloud Practitioner" for example: https://aws.amazon.com/certification/certified-cloud-practitioner/

Author: Yeah, you're right. It's "practitioner". 👍

Now we need to send this function along with the necessary arguments to the HDB. This approach exemplifies a good practice in kdb+: keeping computations as close to the data as possible. Instead of requesting data and then applying a computation to it, we send the computation to the HDB itself so we avoid transmitting unnecessary data over the communication.
```q
t:`sym xgroup hdb(query_prices;4*365;syms)
```
To communicate with the process, we pass a list to the `hdb` handle with the first element being the function and the subsequent elements being the arguments (last 4 years & involved indexes). Once we get our data we finally `xgroup` by index.

> 

Reviewer: I miss closing the connection with `hclose`

Author: I think we could get rid of it, but if you find it relevant, I can add a brief paragraph at the end.

> In our case, we are going to use closing prices to feed the ADF test, so we have to index (`@`) by column **close** from each pair in our table. Additionally, we take (`#`) the last **tr** days of data for both indexes, and finally apply our **fcoint** function to each (`.'`) pair of data lists.

Reviewer: remove (#) explanation, because we dont use it (bug also present in the original post :) )

Author: This is one of the fragments that I wanted to change but I want to discuss with you, there's a comment about this in the original PR.

fcoint:{@[;1]0f^coint[x;y]`}
```

As we saw in the introduction, we are working in a tick architecture environment. This means that, in addition to receiving real-time prices for our indices, this architecture provides a tool to store the closing prices of our indices in our HDB at the end of the day. As obvious, establishing a connection with HDB is required:
```q
hdb:hopen port
```
As can be seen, we assume that the HDB process is listening at a given `port` in the local machine. We just `hopen` a connection with it and get the `hdb` handle.

Then, we define the query that we want the HDB to run. In this case, we declare `rs`, which takes the (n)umber of historic days and the involved (sym)bol(s) for which we want to retrieve data, as arguments.

```q
rs:{[n;syms]select from prices where date within (.z.d-tr;.z.d),sym in syms}
```

> The body of the function might seem pretty familiar to the SQL practicioner. In fact, we are exploiting **qSQL syntax** here, which leverages a syntax similar to SQL but optimised for kdb+. It is also worth noting that `.z.d` represents the current date, so we are interested in retrieving data from the `n` days back from today.

Author: ```suggestion
The body of the function might seem pretty familiar to the SQL practitioner. In fact, we are exploiting **qSQL syntax** here, which leverages a syntax similar to SQL but optimised for kdb+. It is also worth noting that `.z.d` represents the current date, so we are interested in retrieving data from the `n` days back from today.

```

However, this presents **an opportunity for profit** because we know that these assets tend to revert to their shared mean over time. If one asset is **overpriced** and likely to decrease, we may consider **selling it** (going short). Conversely, if an asset is **underpriced** and expected to increase, we may consider **buying it** (going long). That is what we call Pairs Trading.

> 💡 This strategy possesses financial characteristics: our **profitability remains unaffected by the broader market trends**, as our focus lies solely on the disparity between the two assets. It's about relative movements rather than absolute ones; we're indifferent to whether prices are rising or falling. This quality defines it as a **neutral market strategy**.

As you might guess, our next task is to build the model that helps us determine whether an index is actually overpriced or underpriced.

## Determining how to calculate the spreads

> After this initial market assesment, we can move on to coding the actual Pair Trading model that calculates the spreads. The first approach we may try could simply be to subtract them and observe if the difference deviates significantly from zero, considering their scale difference.

Reviewer: I think we should very briefly explain what it means to be overpriced or underpriced since we don't mention it in the entire section, yet it is the reason (I know it's also not present in the original post :) )

Author: Feel free to add whatever change here, I think you'll explain it better than me, but I can try if you want to.

Secondly, the kdb ecosystem provides excellent tools, such as KX Dashboard, which we utilized to produce several graphs, particularly emphasizing real-time visualizations. As a niche platform, kdb cannot compete with Python in terms of ecosystem variety. However, PyKX allows us to overcome this limitation by enabling interaction with the Python ecosystem. This opens up possibilities for leveraging libraries like statsmodels and seaborn, as demonstrated in our examples.

Last but not least, the concision and expressiveness of q is impressive. We only needed to extend the vanilla architecture with less than 40 lines of q code to implement the MS and RPT components. In our view, maintaining 40 lines of code is far preferable to maintaining 400 lines. While becoming proficient with this language takes time, you don't need to master it to start analyzing data. q-SQL is a great entry point for beginners.

However, there are several other benefits of this landscape that we couldn't explore in this article.

> Namely, the performance of these technologies is superb, both in terms of speed and memory footprint. In fact, as stated by KX in one of the articles linked above, kdb+ supports a throughput of 35K ticks per second for a single node. If we scale the solution to several nodes, we should easily accommodate thousands of pairs in the system simultaneously. We leave scaling this solution up as future work.

Reviewer: "35K ticks per second ...", have you tested this? or where did this number come from?

Author: I think it's clearly stated, isn't it? I can supply the link again if you find it necessary...

## Real-time spread calculation

Now that we have selected a pair of cointegrated indices and built a model to calculate their relationships, it's time to formalize its subscription as a real-time component. Once we start receiving data from the TP, we can apply the model to produce the spreads, which will then be sent to the dashboard.

> ![Arch-bottom](resources/general-architecture-bottom.png)

Reviewer 2: ```suggestion

![Arch-bottom](resources/general-architecture-bottom.png)

```

Author: I think it's ok to integrate graphs in the very same paragraph, since they are not independent figures that we are referring to, but if you prefer to keep them in a single one, it's fine.

> Real-time components can manifest their interest for a particular table and for a subset of symbols. As a result from previous steps, we know we are interested on the quotes for SP500 and NASDAQ100:

Reviewer: We should briefly explain why those two hardcoded symbols appear (``` `SP500`NASDAQ100 ```), shouldn't we? I mean, perhaps the reader misses something more automated, right?

Author: Yes, this is one of the problems that I feel while connecting pieces. They seem like independent things while DFA and RPT are completely tied. What we do is weird and difficult to justify, but I'm not sure if we could change it easily. Let me give it a try.

```q
tp"(.u.sub[`quote;`SP500`NASDAQ100])"
```
Assume that `tp` is just a handle to the TP process, similar to `hdb` from previous sections. Basically, what `.u.sub` does is registering the RPT handle in the TP so it can later notify the recently subscribed component about new events. To do so, it assumes that the subscriber has defined an `upd` function:
```q
upd:{.u.pub[`spread;([]time:1#y`time;spread:sp . y`bid)]};
```

> This function essentially takes the current prices of SP500 and NASDAQ100 as input, calculates the spread, and sends it to its subscribers. In this sense, the dashboard should subscribe to the RPT in a similar fashion as the RPT subscribed to the TP.

Reviewer: I don't think this section is bad, but it is a great simplification of what actually happens. I don't mind, I repeat, but I thought it wasn't the objective since the quant might have to work here and has no idea what they are doing

Author: Let me review it so we can add more details in the right direction.

Author: I've added a few things, but I think the new version collects all your original ideas, but in a (hopefully) simpler and easy-to-read way. Do you miss something? We can discuss it.

> This function essentially takes the current prices of SP500 and NASDAQ100 as input, calculates the spread, and sends it to its subscribers. In this sense, the dashboard should subscribe to the RPT in a similar fashion as the RPT subscribed to the TP.

Reviewer: And I would explain in a different way or just remove how the dashboard connects to the RPT, because it's not the same, it's not the same in the sense that we can take advantage of the UI and we don't need to subscribe using 'u.sub'

Author: Sure, we can say that Dashboards does the subscription in the shadows. Will come back to this next week.

> 💡 Signal windows play a pivotal role in implementing Pairs Trading strategies. They serve as indicators for determining when to execute buy and sell actions, acting as arbitrary thresholds that guide our algorithm's decision-making process. These windows are derived from the variance of our data, representing a static variance assumption due to our consideration of a time-independent cointegrated series.

## Conclusion

In this post, we have provided a comprehensive overview of the implementation of the Pairs Trading strategy in KDB+/Q, contextualized within the Tick Architecture. Here are some key takeaways:
* The Tick Architecture allows us to handle both historical and real-time data.
* By leveraging historical data, we were able to easily identify cointegrated pairs, reusing libraries from the Python ecosystem via PyKX when needed.
* Q is very expressive and the implementation of the Linear Regression logic for producing the spread model is straightforward.
* Integrating a real-time component and connecting it with a dashboard is simple and efficient.

> More generally, and although we couldn't delve into all the details in this post, we'd like to emphasize the three major selling points of KDB+/Q. First, it can process large amounts of data in a very short time with a small memory footprint, allowing us to monitor thousands of pairs simultaneously. Secondly, Q code is highly concise, enabling us to implement all the components in the diagram in less than 100 lines of code. Finally, the technology is highly flexible, allowing us to easily adapt to other scenarios beyond Pairs Trading.

Reviewer 2: ```suggestion
More generally, and although we couldn't get into all the details in this post, we'd like to emphasize the three major selling points of KDB+/Q. First, it can process large amounts of data in a very short time with a small memory footprint, allowing us to monitor thousands of pairs simultaneously. Secondly, Q code is highly concise, enabling us to implement all the components in the diagram in less than 100 lines of code. Finally, the technology is highly flexible, allowing us to easily adapt to other scenarios beyond Pairs Trading.

```

Author: Ok, you don't like _delve_ at all 😆
